### ISSUE TITLE
[FEATURE] Support LSP4IJ to be available on all IntelliJ-based IDEs

### ISSUE THREAD

**Issue by antruongnguyen** [cap-js/cds-intellij#82]
### Description

Currently, the plugin is only available for the **paid** versions of IntelliJ IDEs. The reason is that the plugin uses the LSP API, which is only available in commercial IntelliJ IDEs. This is why the plugin does not work with the free variants (such as IntelliJ Community).

**LSP4IJ** is a free and open-source Language Server Protocol (LSP) client compatible with all flavors of IntelliJ. The LSP4IJ plugin is available [here](https://plugins.jetbrains.com/plugin/23257-lsp4ij), and its source code is located [here](https://github.com/redhat-developer/lsp4ij).

### Suggested Solution

Please support **[LSP4IJ](https://plugins.jetbrains.com/plugin/23257-lsp4ij)** in order to replace [IntelliJ LSP API](https://plugins.jetbrains.com/docs/intellij/language-server-protocol.html)

Source code: https://github.com/redhat-developer/lsp4ij

**Comment by manwithafox** [comment]
@antruongnguyen Thank you for this hint. Actually in the past we had an (SAP internal) IJ plugin based on another LSP client lib. But that had a pretty clumsy integration. Far away from production quality.

The current approach gives us more trust as IJ itself provides the client side integration. Though, as you spotted, they do not support community editions, AND, while continuously improving, still lack many feature.

Taking a look into the source of LSP4IJ shows that they already have very broad request support and they are pacing fast - just a year since initial release.

I did a short manual configuration to try it out:

In `Settings->Languages&Frameworks->Language Servers` I created a new entry `CDS` with command `node /path/to/@sap/cds-lsp/dist/main.js  --stdio`. 
In the `Mappings` tab added an entry for file pattern `*.cds` with lang id `cds`, left `Configuration` tab empty.
Et voilà. It works already pretty nicely. Though there are some quirks we need to investigate. 

E.g. 

- Alt-F7 Find Usages is not supported, but a proprietary GoTo->LSP references with special dialog. 
- Command-Hover highlights the wrong scope.
- ... to be extended

**Comment by angelozerr** [comment]
> Alt-F7 Find Usages is not supported, but a proprietary GoTo->LSP references with special dialog.

@antruongnguyen please note that LSP4IJ have a luck to have the contributor @SCWells72 which is working hard to have standard IDE actions like https://github.com/redhat-developer/lsp4ij/pull/637

> Command-Hover highlights the wrong scope.

Could you create an issue with a detailed info please.

> ... to be extended

Don'thesitate to create any issues.

**Comment by angelozerr** [comment]
> @antruongnguyen Thank you for this hint. Actually in the past we had an (SAP internal) IJ plugin based on another LSP client lib. But that had a pretty clumsy integration. Far away from production quality.
> 
> The current approach gives us more trust as IJ itself provides the client side integration. Though, as you spotted, they do not support community editions, AND, while continuously improving, still lack many feature.
> 
> Taking a look into the source of LSP4IJ shows that they already have very broad request support and they are pacing fast - just a year since initial release.
> 
> I did a short manual configuration to try it out:

Please note that you can import / export the configuration that you have explained below. See https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserDefinedLanguageServer.md#custom-template

> 
> In `Settings->Languages&Frameworks->Language Servers` I created a new entry `CDS` with command `node /path/to/@sap/cds-lsp/dist/main.js  --stdio`. 
> In the `Mappings` tab added an entry for file pattern `*.cds` with lang id `cds`, left `Configuration` tab empty.
> Et voilà. It works already pretty nicely. Though there are some quirks we need to investigate. 
> 
> E.g. 
> 
> - Alt-F7 Find Usages is not supported, but a proprietary GoTo->LSP references with special dialog. 
> - Command-Hover highlights the wrong scope.
> - ... to be extended
> 
>

**Comment by manwithafox** [comment]
Just a short notice: I'm currently working on this... : )

**Comment by manwithafox** [comment]
> [@manwithafox](https://github.com/manwithafox?rgh-link-date=2025-03-11T13%3A25%3A49.000Z) I have noticed that you are using now LSP4IJ, it is fantastic!
> 
> Could you please create a PR to LSP4IJ to add it to [redhat-developer/lsp4ij#who-is-using-lsp4ij](https://github.com/redhat-developer/lsp4ij?tab=readme-ov-file&rgh-link-date=2025-03-11T13%3A25%3A49.000Z#who-is-using-lsp4ij) please
> 
> And if you like LSP4IJ could you please add a review [plugins.jetbrains.com/plugin/23257-lsp4ij/reviews](https://plugins.jetbrains.com/plugin/23257-lsp4ij/reviews) to help us to promote LSP4IJ
> 
> Thanks!

Once we release this to the Marketplace I will create the PR. Thanks

**Comment by tim-sh** [comment]
[Version 2.0.0 is out](https://plugins.jetbrains.com/plugin/25209-sap-cds-language-support/versions/stable/835266) and with it, support for IDEA Community.[^1]

[^1]: For the record: 'Compatible with IntelliJ IDEA (Ultimate, Community), Android Studio, AppCode, Aqua, CLion, Code With Me Guest, DataGrip, DataSpell, GoLand, JetBrains Client, JetBrains Gateway, MPS, PhpStorm, PyCharm, Rider, RubyMine, RustRover, WebStorm, Writerside.'

**Comment by manwithafox** [comment]
Indeed. But should be irrelevant for us if they need to include their LSP client into the Community (unified) edition. 
LSP4IJ is much more advanced than their client.
